moto/iam/access_control.py

```diff
@@ -258,9 +263,41 @@ def check_action_permitted(self, resource: str) -> None:
             elif permission_result == PermissionResult.PERMITTED:
                 permitted = True
 
+        if self._is_assuming_role_operation(resource):
+            permitted = self._check_role_trust_relationship(resource)
+
         if not permitted:
             self._raise_access_denied()
 
+    def _is_assuming_role_operation(self, resource_arn: str) -> bool:
+        if ":role" not in resource_arn.lower():
+            return False
+
+        if not self.backend.has_role_by_arn(resource_arn):
+            return False
+
+        self.backend.get_role_by_arn(resource_arn)
+        return self._action in REQUIRE_RESOURCE_ACCESS_POLICIES_CHECK
```

Comment:
This loops through all the IAM roles twice - first to see if it exists, and then again to return it. It would be faster to only run through the loop once, by only using the existing `get_role_by_arn`-method:

```suggestion
        try:
            self.backend.get_role_by_arn(resource_arn)
            return self._action in REQUIRE_RESOURCE_ACCESS_POLICIES_CHECK
        except IAMNotFoundException:
            return False
```

Reply:
Fixed